packages/@lwc/engine-core/src/framework/rendering.ts

```diff
@@ -588,7 +588,12 @@ function updateDynamicChildren(parentElm: Node, oldCh: VNodes, newCh: VNodes) {
                         newStartVnode.hook.insert(newStartVnode, parentElm, oldStartVnode.elm!);
                     } else {
                         patchVnode(elmToMove, newStartVnode);
-                        oldCh[idxInOld] = undefined as any;
+                        // Delete the old child, but copy the array since it is read-only.
```

Comment:
my intent is to remove (or minimize) this cloning from the codepath, I have some questions:
1. is it that we declared [VNodes readonly](https://github.com/salesforce/lwc/pull/2618/files#diff-8079a416c73a132a9a93647a3f1fe5f856e79c5dda656da2cd6c333c0909ffa8R21) for convenience, or that it must be readonly?
2. if it is for convenience, can we redeclare the type in `updateDynamicChildren` to receive a Writable<VNodes>?
3. if it is a must (or we want to be strict with the type), would it help just to clone the array once, instead of multiple times?

My thought process is that we may need to move multiple nodes via this codepath, and all this cloning may happen multiple times, when it really needs to happen only once.

Reply:
1 - Yes it is required. Vnodes is the children array, and it may be shared (due to empty arrays currently, but in the future I'd like static children to be completely shared too).

2 - Hm, the thing is that this code path is _very_ uncommon (it wasn't even covered in any of our Karma tests until I added a test), so we don't want to clone the array every time the function is called. We can certainly clone once (when the codepath is hit), but I don't know if it would actually improve perf in many scenarios, since this scenario is already so rare.

I'll open a separate issue to explore the "only clone once," since I think we need to explore whether it actually improves perf and by how much.